File: 05-07 Multivariate Outliers SVMs Ensembles/06_svms.py
"""
Chapter 6: Support Vector Machines
"""

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import roc_auc_score, log_loss, confusion_matrix

d = pd.read_csv("WBCdata_clean.csv")

seed = 14

x_train, x_test, y_train, y_test = train_test_split(d.iloc[:, 3:], d.diagnosis,
                                                    test_size=0.33, random_state=seed,
                                                    shuffle=True, stratify=d.diagnosis)


#Setting up model parameters
boundary_smoothness = 1/len(d.columns[3:])
svc_lin_model = SVC(kernel='linear', random_state=seed)
svc_pol_model = SVC(kernel='poly', degree=3, gamma=boundary_smoothness, random_state=seed)
svc_rbf_model = SVC(kernel='rbf', gamma=boundary_smoothness, random_state=seed)
svc_sig_model = SVC(kernel='sigmoid', gamma=boundary_smoothness, random_state=seed)

#Training models
svc_lin_model.fit(x_train, y_train)
svc_pol_model.fit(x_train, y_train)
svc_rbf_model.fit(x_train, y_train)
svc_sig_model.fit(x_train, y_train)

#Predictions
svc_lin_preds = svc_lin_model.predict(x_test)
svc_pol_preds = svc_pol_model.predict(x_test)
svc_rbf_preds = svc_rbf_model.predict(x_test)
svc_sig_preds = svc_sig_model.predict(x_test)

#Evaluation
print("Linear SVC AUC", roc_auc_score(y_test, svc_lin_preds))
print("Polynomial SVC AUC", roc_auc_score(y_test, svc_pol_preds))
print("Radial Basis SVC AUC", roc_auc_score(y_test, svc_rbf_preds))
print("Sigmoid SVC AUC", roc_auc_score(y_test, svc_sig_preds))
print("\n")
print("Linear SVC Log Loss", log_loss(y_test, svc_lin_preds))
print("Polynomial SVC Log Loss", log_loss(y_test, svc_pol_preds))
print("Radial Basis SVC Log Loss", log_loss(y_test, svc_rbf_preds))
print("Sigmoid SVC Log Loss", log_loss(y_test, svc_sig_preds))
print("\n")
print("Confusion Matrix Format: \n[TN, FP]\n[FN, TP]")
print("Linear SVC Confusion Matrix \n", confusion_matrix(y_test, svc_lin_preds))
print("Polynomial SVC Confusion Matrix \n", confusion_matrix(y_test, svc_pol_preds))
print("Radial Basis SVC Confusion Matrix \n", confusion_matrix(y_test, svc_rbf_preds))
print("Sigmoid SVC Log Confusion Matrix \n", confusion_matrix(y_test, svc_sig_preds))

File: 05-07 Multivariate Outliers SVMs Ensembles/06_svms.R

setwd("dsilt-ml-code/05-07 Multivariate Outliers SVMs Ensembles")

library(caret)
library(e1071)
library(MLmetrics)

d <- read.csv("WBCdata_clean.csv", header=TRUE)
head(d)

seed <- 14

train_ind <- createDataPartition(y=d$diagnosis, p=0.66, list=F)
train <- d[train_ind, 3:12]
test <- d[-train_ind, 3:12]

#Setting up model parameters and training models
boundary_smoothness <- 1/ncol(train)
svc_lin_model <- svm(diagnosis ~ ., data=train, 
                     scale=TRUE, type='C-classification',
                     kernel='linear', seed=seed)
svc_pol_model <- svm(diagnosis ~ ., data=train, 
                     scale=TRUE, type='C-classification',
                     kernel='polynomial', degree=3,
                     gamma=boundary_smoothness,
                     seed=seed)
svc_rbf_model <- svm(diagnosis ~ ., data=train, 
                     scale=TRUE, type='C-classification',
                     kernel='radial', 
                     gamma=boundary_smoothness,
                     seed=seed)
svc_sig_model <- svm(diagnosis ~ ., data=train, 
                     scale=TRUE, type='C-classification',
                     kernel='sigmoid', 
                     gamma=boundary_smoothness,
                     seed=seed)

#Predictions
svc_lin_preds <- as.integer(predict(svc_lin_model, test[,2:10]))
svc_pol_preds <- as.integer(predict(svc_pol_model, test[,2:10]))
svc_rbf_preds <- as.integer(predict(svc_rbf_model, test[,2:10]))
svc_sig_preds <- as.integer(predict(svc_sig_model, test[,2:10]))

#Evaluation
print(paste("Linear SVC AUC", AUC(svc_lin_preds, test$diagnosis)))
print(paste("Polynomial SVC AUC", AUC(svc_pol_preds, test$diagnosis)))
print(paste("Radial Basis SVC AUC", AUC(svc_rbf_preds, test$diagnosis)))
print(paste("Sigmoid SVC AUC", AUC(svc_sig_preds, test$diagnosis)))
print(paste(" "))
print(paste("Linear SVC Log Loss", LogLoss(svc_lin_preds, test$diagnosis)))
print(paste("Polynomial SVC Log Loss", LogLoss(svc_pol_preds, test$diagnosis)))
print(paste("Radial Basis SVC Log Loss", LogLoss(svc_rbf_preds, test$diagnosis)))
print(paste("Sigmoid SVC Log Loss", LogLoss(svc_sig_preds, test$diagnosis)))
print(paste(" "))
print("Linear SVC Confusion Matrix")
ConfusionMatrix(svc_lin_preds, test$diagnosis)
print("Polynomial SVC Confusion Matrix")
ConfusionMatrix(svc_pol_preds, test$diagnosis)
print("Radial Basis SVC Confusion Matrix")
ConfusionMatrix(svc_rbf_preds, test$diagnosis)
print("Sigmoid SVC Log Confusion Matrix")
ConfusionMatrix(svc_sig_preds, test$diagnosis)
